### PR TITLE
[SPARK-28151][SQL] Fix MsSqlServerDialect Byte/Short/Float type mappings ( DRAFT)

### PR DESCRIPTION
**Pull request by shivsood**
Abandoning this PR. Will create new one that have specific fixes
1. PR for Float Type and ShortType Fix 
2. PR for ByteType Fix.

Fix for ByteType, ShortType and FloatTypes are not correctly mapped for read/write of SQLServer tables
###ByteType issue
Writing dataframe with column type BYTETYPE fails when using JDBC connector for SQL Server. Append and Read of tables also fail. The problem is due 

1. (Write path) Incorrect mapping of BYTETYPE in getCommonJDBCType() in jdbcutils.scala where BYTETYPE gets mapped to BYTE text. It should be mapped to TINYINT
case ByteType => Option(JdbcType("BYTE", java.sql.Types.TINYINT))

In getCatalystType() ( JDBC to Catalyst type mapping) TINYINT is mapped to INTEGER, while it should be mapped to BYTETYPE. Mapping to integer is ok from the point of view of upcasting, but will lead to 4 byte allocation rather than 1 byte for BYTETYPE.

2. (read path) Read path ends up calling makeGetter(dt: DataType, metadata: Metadata). The function sets the value in RDD row. The value is set per the data type. Here there is no mapping for BYTETYPE and thus results will result in an error when getCatalystType() is fixed.

Note : These issues were found when reading/writing with SQLServer. Will be submitting a PR soon to fix these mappings in MSSQLServerDialect.

Error seen when writing table

(JDBC Write failed,com.microsoft.sqlserver.jdbc.SQLServerException: Column, parameter, or variable #2: Cannot find data type BYTE.)
com.microsoft.sqlserver.jdbc.SQLServerException: Column, parameter, or variable #2: Cannot find data type BYTE.
com.microsoft.sqlserver.jdbc.SQLServerException.makeFromDatabaseError(SQLServerException.java:254)
com.microsoft.sqlserver.jdbc.SQLServerStatement.getNextResult(SQLServerStatement.java:1608)
com.microsoft.sqlserver.jdbc.SQLServerStatement.doExecuteStatement(SQLServerStatement.java:859)
 ..

###ShortType and FloatType issue
ShortType and FloatTypes are not correctly mapped to right JDBC types when using JDBC connector. This results in tables and spark data frame being created with unintended types.

Some example issue

Write from df with column type results in a SQL table of with column type as INTEGER as opposed to SMALLINT. Thus a larger table that expected.
read results in a dataframe with type INTEGER as opposed to ShortType

FloatTypes have a issue with read path. In the write path Spark data type 'FloatType' is correctly mapped to JDBC equivalent data type 'Real'. But in the read path when JDBC data types need to be converted to Catalyst data types ( getCatalystType) 'Real' gets incorrectly gets mapped to 'DoubleType' rather than 'FloatType'.

## What changes were proposed in this pull request?
(MsSqlServerDialect.scala ) BYTETYPE  is mapped to "TINYINT", SHORTTYPE is mapped to small int
(JdbcUtils.scala) reading ByteTYPE is added to makeGetter to enable reading after ByteType is translated to TINYINT.

## How was this patch tested?
Unit test - Added and passed.
Integration Test - Tested end to end with SQLServer using JDBC connector.

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)
(If this patch involves UI changes, please attach a screenshot; otherwise, remove this)

Please review https://spark.apache.org/contributing.html before opening a pull request.

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala [hunk 1]
```diff
@@ -895,6 +895,24 @@ class JDBCSuite extends QueryTest
...
     assert(msSqlServerDialect.getJDBCType(BinaryType).map(_.databaseTypeDefinition).get ==
       "VARBINARY(MAX)")
+    assert(msSqlServerDialect.getJDBCType(ByteType).map(_.databaseTypeDefinition).get ==
+      "TINYINT")
+
+    assert(msSqlServerDialect.getJDBCType(ShortType).map(_.databaseTypeDefinition).get ==
+      "SMALLINT")
+  }
+
+  test("MsSqlServerDialect catalyst type mapping") {
```
Comment (wangyum) [inline sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala]: It is best to add tests in [`MsSqlServerIntegrationSuite`](https://github.com/apache/spark/blob/master/external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MsSqlServerIntegrationSuite.scala) as well.

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/MsSqlServerDialect.scala [hunk 2]
```diff
@@ -30,7 +30,13 @@ private object MsSqlServerDialect extends JdbcDialect {
       // String is recommend by Microsoft SQL Server for datetimeoffset types in non-MS clients
       Option(StringType)
     } else {
-      None
+      val answer = sqlType match {
```
Comment (srowen) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/MsSqlServerDialect.scala]: `Option(x)` should be `Some(x)` here, though both will work. Same below. You can remove `answer` as redundant.

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/MsSqlServerDialect.scala [hunk 3]
```diff
@@ -39,6 +45,8 @@ private object MsSqlServerDialect extends JdbcDialect {
     case StringType => Some(JdbcType("NVARCHAR(MAX)", java.sql.Types.NVARCHAR))
     case BooleanType => Some(JdbcType("BIT", java.sql.Types.BIT))
     case BinaryType => Some(JdbcType("VARBINARY(MAX)", java.sql.Types.VARBINARY))
+    case ByteType => Option(JdbcType("TINYINT", java.sql.Types.TINYINT))
+    case ShortType => Option(JdbcType("SMALLINT", java.sql.Types.SMALLINT))
```
Comment (srowen) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/MsSqlServerDialect.scala]: Does FloatType need to map back to REAL or does that already happen?

**Review by srowen — COMMENTED** [review]
I don't know SQL Server, so can't really evaluate it, but seems like a plausible change.

File: sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala (hunk 1 above)
Comment (shivsood) [reply sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala]: @wangyum Thanks for pointing me to MsSqlServerIntegration.  Agree that's a good place to add these test. Looks like these test spin up a SQL server for test. Is this run in CI? Any pointers to how to run these test.

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/MsSqlServerDialect.scala (hunk 3 above)
Comment (shivsood) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/MsSqlServerDialect.scala]: Yes,  FloatType is correctly mapped to REAL in def getCommonJDBCType(dt: DataType)

File: sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala (hunk 1 above)
Comment (wangyum) [reply sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala]:
```
./build/mvn install -DskipTests
./build/mvn test -Pdocker-integration-tests -pl :spark-docker-integration-tests_2.12
```

http://spark.apache.org/docs/latest/building-spark.html#running-docker-based-integration-test-suites

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/MsSqlServerDialect.scala (hunk 2 above)
Comment (shivsood) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/MsSqlServerDialect.scala]: Thanks. Will make a fix. Change to Some(x) will avoid the null check which is redundant in this case.

File: sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala (hunk 1 above)
Comment (shivsood) [reply sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala]: Thanks

Comment (shivsood) [reply sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala]: @wangyum Thanks for this comment. Integration test brought to light another issue that i have now fixed in JDBCutils.scala (Byte was being written as Int, thus Byte with value 255 was interpreted as -1, resulting in a JDBC exception).

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/MsSqlServerDialect.scala (hunk 2 above)
Comment (shivsood) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/MsSqlServerDialect.scala]: Fixed now.

**Comment by dongjoon-hyun** [comment]
ok to test

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
Hi, @shivsood . Thank you for your contribution.
Could you run `dev/scalastyle` in your environment and fix all errors?

Currently, this PR has the followings.
```
========================================================================
Running Scala style checks
========================================================================
[info] Checking Scala style using SBT with these profiles:  -Phadoop-2.7 -Pkubernetes -Phive-thriftserver -Phadoop-cloud -Pkinesis-asl -Pyarn -Pspark-ganglia-lgpl -Phive -Pmesos
Scalastyle checks failed at following occurrences:
[error] /home/jenkins/workspace/SparkPullRequestBuilder@2/sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala:909:68: No space after token ,
[error] /home/jenkins/workspace/SparkPullRequestBuilder@2/sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala:909:78: No space after token ,
[error] /home/jenkins/workspace/SparkPullRequestBuilder@2/sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala:909:80: No space after token ,
[error] /home/jenkins/workspace/SparkPullRequestBuilder@2/sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala:911:69: No space after token ,
[error] /home/jenkins/workspace/SparkPullRequestBuilder@2/sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala:911:80: No space after token ,
[error] /home/jenkins/workspace/SparkPullRequestBuilder@2/sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala:911:82: No space after token ,
[error] /home/jenkins/workspace/SparkPullRequestBuilder@2/sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala:913:65: No space after token ,
[error] /home/jenkins/workspace/SparkPullRequestBuilder@2/sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala:913:72: No space after token ,
[error] /home/jenkins/workspace/SparkPullRequestBuilder@2/sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala:913:74: No space after token ,
[error] Total time: 27 s, completed Jul 5, 2019 1:08:48 AM
```

File: sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala [hunk 4]
```diff
@@ -895,6 +895,24 @@ class JDBCSuite extends QueryTest
...
+    val msSqlServerDialect = JdbcDialects.get("jdbc:sqlserver")
+    val metadata = new MetadataBuilder().putLong("scale", 1)
+
+    assert(msSqlServerDialect.getCatalystType(java.sql.Types.TINYINT,"TINYINT",1,metadata).get ==
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala]: We need spaces among parameters.

File: sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala [hunk 5]
```diff
@@ -895,6 +895,24 @@ class JDBCSuite extends QueryTest
...
+          ByteType)
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala]: indentation?

File: sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala (hunk 4 above)
Comment (shivsood) [reply sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala]: Thanks dongjoon. I have resolved these.

File: sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala (hunk 5 above)
Comment (shivsood) [reply sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala]: Resolved. Sorry, about this oversight.

**Comment by shivsood** [comment]
> **[Test build #107364 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107364/testReport)** for PR 24969 at commit [`8a7d960`](https://github.com/apache/spark/commit/8a7d960833799ba4a20e90b21e564b1d8a61f424).
> 
>     * This patch **fails to build**.
> 
>     * This patch merges cleanly.
> 
>     * This patch adds no public classes.

Not clear why this failure. Can someone guide please.

[info] Done packaging.
java.lang.RuntimeException: 1 fatal warnings
	at scala.sys.package$.error(package.scala:27)
	at SparkBuild$$anonfun$sharedSettings$21.apply(SparkBuild.scala:315)
	at SparkBuild$$anonfun$sharedSettings$21.apply(SparkBuild.scala:284)
	at scala.Function1$$anonfun$compose$1.apply(Function1.scala:47)
	at sbt.$tilde$greater$$anonfun$$u2219$1.apply(TypeFunctions.scala:40)
	at sbt.std.Transform$$anon$4.work(System.scala:63)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
	at sbt.Execute.work(Execute.scala:237)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestrictions.scala:159)
	at sbt.CompletionService$$anon$2.call(CompletionService.scala:28)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
[error] (sql/compile:compile) 1 fatal warnings
[error] Total time: 161 s, completed Jul 8, 2019 10:57:09 AM
[error] running /home/jenkins/workspace/SparkPullRequestBuilder/build/sbt -Phadoop-2.7 -Pkubernetes -Phive-thriftserver -Phadoop-cloud -Pkinesis-asl -Pyarn -Pspark-ganglia-lgpl -Phive -Pmesos test:package streaming-kinesis-asl-assembly/assembly ; received return code 1
Attempting to post to Github...
 > Post successful.
Build step 'Execute shell' marked build as failure
Archiving artifacts
Recording test results
ERROR: Step ?Publish JUnit test result report? failed: No test report files were found. Configuration error?
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/107364/
Test FAILed.
Finished: FAILURE

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala [hunk 6]
```diff
@@ -797,6 +797,7 @@ object JdbcUtils extends Logging {
       tableSchema: StructType,
       customSchema: String,
       nameEquality: Resolver): StructType = {
+
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: Let's remove this addition.

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/MsSqlServerDialect.scala [hunk 7]
```diff
@@ -30,7 +30,12 @@ private object MsSqlServerDialect extends JdbcDialect {
       // String is recommend by Microsoft SQL Server for datetimeoffset types in non-MS clients
       Option(StringType)
     } else {
-      None
+        sqlType match {
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/MsSqlServerDialect.scala]: Indentation?

File: sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala [hunk 8]
```diff
@@ -895,6 +895,24 @@ class JDBCSuite extends QueryTest
       "BIT")
     assert(msSqlServerDialect.getJDBCType(BinaryType).map(_.databaseTypeDefinition).get ==
       "VARBINARY(MAX)")
+    assert(msSqlServerDialect.getJDBCType(ByteType).map(_.databaseTypeDefinition).get ==
+      "TINYINT")
+
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala]: Let's remove this empty line.

File: sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala [hunk 9]
```diff
@@ -895,6 +895,24 @@ class JDBCSuite extends QueryTest
...
+    val msSqlServerDialect = JdbcDialects.get("jdbc:sqlserver")
+    val metadata = new MetadataBuilder().putLong("scale", 1)
+
+    assert(msSqlServerDialect.getCatalystType(java.sql.Types.TINYINT, "TINYINT", 1,
+      metadata).get == ByteType)
+    assert(msSqlServerDialect.getCatalystType(java.sql.Types.SMALLINT, "SMALLINT", 1,
+      metadata).get == ShortType)
+    assert(msSqlServerDialect.getCatalystType(java.sql.Types.REAL, "REAL", 1,
+      metadata).get == FloatType)
+
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala]: Let's remove this empty line.

File: sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala (hunk 1 above)
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala]: Since this is a bug fix, please use `SPARK-28151` prefix to this test case.
```scala
-  test("MsSqlServerDialect catalyst type mapping") {
+  test("SPARK-28151 MsSqlServerDialect catalyst type mapping") {
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala [hunk 10]
```diff
@@ -550,7 +550,7 @@ object JdbcUtils extends Logging {
 
     case ByteType =>
       (stmt: PreparedStatement, row: Row, pos: Int) =>
-        stmt.setInt(pos + 1, row.getByte(pos))
+        stmt.setByte(pos + 1, row.getByte(pos))
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: This is orthogonal to MySqlServerDialect. Can we proceed this PR without?

Comment (dongjoon-hyun) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: Or, we can proceed this first as another PR with a proper regression test.

**Comment by dongjoon-hyun** [comment]
@shivsood . `JdbcUtils.scala` change should be handled in a separate PR.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala (hunk 10 above)
Comment (shivsood) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: @dongjoon-hyun JDBCUtils.scala fix is required for completeness of the ByteType Fix. Without this fix Integrations test  in "external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MsSqlServerIntegrationSuite.scala" will fail. 

I dont think the "MsSqlServerIntegrationSuite.scala"  integration tests are running in CI/CD. I was running these locally following a review suggestion. Is that's true, i can still remove this fix and submit this as part of another PR.

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/MsSqlServerDialect.scala (hunk 7 above)
Comment (shivsood) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/MsSqlServerDialect.scala]: sorry about this. Resolved now.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala (hunk 10 above)
Comment (dongjoon-hyun) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: Then, you need to follow the second option I gave.
> Or, we can proceed this first as another PR with a proper regression test.

Comment (shivsood) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: @dongjoon-hyun  Can u please elaborate on proper regression test. I have test cases added/fixed for ByteType in JDBCSuite.scala and MsSqlServerIntegrationSuite.scala. Are you expecting some other test? Can u please point me to type of test.

Another way to cleanly split this PR would be as follows. 
1. Remove ByteType fix from this PR and jdbcutils.scala fix also get removed. This PR will only have fix for FloatType and Short.
2. 2nd PR with complete ByteType fix with more elaborate integration test.

I prefer this for better seperation of concern in each of the PRs , while still retaining the completeness. Let me know what you think.

Thanks for your patience and very helpful guidance.

Comment (dongjoon-hyun) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: What I meant is like your 2nd option.
> 2nd PR with complete ByteType fix with more elaborate integration test.

Here, the proper regression test means a test causing failure for the other Dialects. (e.g. PostgreSQL/MySQL/...) because `JdbcUtils` is shared across Dialects.

Comment (shivsood) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: @dongjoon-hyun  Would create 2 PRs then. PR1 ( this PR) i will remove ByteType and JDBCUtilis fix. Will create PR2 for ByteType and JDBCUtilis fix with full integration test. PR2 will take time as i have to investigate how to test for other dialects.

Comment (dongjoon-hyun) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala]: Yep. Please try as you want. Then, ping me later when the PR is ready.

**Comment by srowen** [comment]
@shivsood do you want to update this PR accordingly?

**Comment by shivsood** [comment]
> @shivsood do you want to update this PR accordingly?
'll abandon this PR and start with a new/clean one. Have update the summary.

**Comment by srowen** [comment]
@shivsood did you open a new PR?
This one looked pretty fine overall minus moving out some JdbcUtils changes. I don't want to lose your improvement.

**Comment by dongjoon-hyun** [comment]
Yes. We are working on https://github.com/apache/spark/pull/25146 .

**Comment by shivsood** [comment]
@srowen Working on #25146 for FloatType and ShortType change. Will raise a PR for ByteType change this week ( It need more testing and i am finding my way thru adding E2E regression test)

**Comment by nkronenfeld** [comment]
Was a PR ever submitted on the Byte/TinyInt problem? If not, does anyone know why? We are still getting burned by this.

**Comment by frgomes** [comment]
@shivsood Hi, what is the PR for ByteType, please?

**Comment by frgomes** [comment]
@dongjoon-hyun  :: I'm getting this exception below which makes me consider that maybe your fix was not applied.

    com.microsoft.sqlserver.jdbc.SQLServerException: Column, parameter, or variable #7: Cannot find data type BYTE.

Was your fix for ByteType ever submitted/applied? To which branch, please?

Thanks

**Comment by srowen** [comment]
https://github.com/apache/spark/pull/25146 ? see above

**Comment by frgomes** [comment]
@srowen : Thanks

**Comment by frgomes** [comment]
@srowen :: #25146 only covers ShortType (smallint) and FloatType (real). TinyType (tinyint) seems to be covered here and only here. However, I'm facing an issue which makes be believe that the fix somehow is not implemented in the runtime I'm using.

**Comment by frgomes** [comment]
@srowen : Feedback: As suspected, the issue was between the keyboard and the chair. Maybe I was not casting properly, or maybe a typo or maybe some equality test somewhere... whatever... when I cast like this ``.cast(pyspark.sql.types.ByteType())``, it works.
